fix and cleanup audio device lists

antis81 opened the pull request:
- [x] cleanup audio device list creation and usage in qTox
- [x] fix selection of default audio device on initialization
- [x] fix initialization, when no device is selected (do not initialize)
- [x] actually disable an audio device without losing the name information
- [x] **WON'T FIX**: ~~fix unicode conversion on Windows (probably only affects pre Win10 versions ??)~~
  - String encoding in OpenAL is undocumented and can lead to incorrect device names on older platforms - especially Windows. With RtAudio, this problem is solved. IMO it is not worth the effort.

TheNain38 commented on src/audio/audio.cpp:
```diff
@@ -524,20 +536,40 @@ bool Audio::isOutputReady()
+        while (*pDeviceList)
+        {
+            int len = strlen(pDeviceList);
+            list <<  QString::fromUtf8(pDeviceList, len);
```
There are two spaces, instead of one

antis81 commented on src/widget/form/settings/avform.cpp:
```diff
@@ -321,52 +321,30 @@ void AVForm::getVideoDevices()
-            pDeviceList += len+1;
-        }
-    }
+    bodyUI->inDevCombobox->addItems(Audio::inDeviceNames());
```
The `deviceNames` list should be used here and initialized with Audio::inDevNames. :) Sorry, will clean this up later today.

zetok commented:
@antis81 um, there is now used commit message format, which will allow to automatically generate changelog for the next, and following releases :)
https://github.com/tux3/qTox/blob/master/CONTRIBUTING.md#-git-commit-guidelines

antis81 commented:
@zetok Thanks, I'll change the message.

antis81 committed:
refactor(audio): cleanup initialization of audio device lists

* move private code
* fix unsigned/signed conversion warnings

antis81 committed:
refactor(audio): cleanup initialization of audio device lists

API:
* open default in/out device, when device name is empty
* move privatized code
* fix unsigned/signed conversion warnings

UI (settings):
* refactor: select audio devices by combobox index

zetok commented:
@antis81 um, you've said in #3125 that this PR might not be merged because it's still WIP – if the code is good for master, it should be merged, even if it doesn't fix all the bugs that are out there ;)

But due to "WIP" label this PR seemed like something that isn't finished yet, so e.g. I didn't even test it, and I'd guess that it goes in similar way for everyone else who would be interested in it getting merged.

Meanwhile, it got conflicts with master :(

Dunno, if it's ready for merge, even if it doesn't fix all the things that don't work too well, but if it improves stuff, or makes stuff ready for improvements, IMO "WIP" should be removed (after PR gets rebased on newest master).

antis81 commented:
> @antis81 um, you've said in #3125 that this PR might not be merged because it's still WIP…

No, I said it probably will not be merged, because we plan to switch API. No problem. I have some local changes lingering around, which solve the conflicts and add some more value. I'll finish those changes and push, so it can actually be merged without conflicts. As always, job/life situation keeps changing in IT business and I have to prioritize.

antis81 commented:
@zetok No more conflicts. I'll push some more changes, which correct enable/disable audio output without mixing it with device names. After that, we are ready to merge.

antis81 committed:
fix(audio): actually disable the audio in/out device in settings, when selected

As esecially the "Disabled" text is translated, the audio device will change from "disabled" to "default", just by changing the language.

In contrast to video devices, an audio device is either always available, or sound will be disabled. So "Disabled" is the correct term to use here.

antis81 commented:
@zetok Guess we're fine here. Some review would be nice, but not urgently required. FYI: As OpenAL's original text encoding concerning device names, the last topic cannot be fixed easily and requires deep investigation of OpenAL sources. This is the actual reason, I didn't merge this. But, as we decided to move from OpenAL and use a library more friendly to utf-8, I don't think it is worth the effort. If someone likes to take the time figuring out the encoding, feel free to fix it…

zetok commented:
@antis81 

> No, I said it probably will not be merged, because we plan to switch API. No problem. (…)

Looks like I've misunderstood >.<

# 

on 765eac9 I got some segfault: https://gist.github.com/zetok/f8e3ab215a355e0f1c46fdcb3b5c8fdd

Happened when I clicked on `Exit` in tray menu after
- starting audio conversation
- starting group audio call
- switching between devices multiple times in settings (also re-scanning devices after adding new one)
- ended audio conversation
- without ending group call tried to `Exit`, while qTox instance had `Audio/Video` settings tab open

backtrace looks a bit weird, could it be something in toxav ?

antis81 commented:
@zetok Thanks for trying with a user's viewpoint. Are you sure the segfault only happens with the commit you stated above? I absolutely couldn't find anything related. Even the corrected boolean return value from `Audio::initInput` only returns correctly `false` now, when device is disabled - nothing bad happening here.

> backtrace looks a bit weird, could it be something in toxav ?

Probably. When looking at the backtrace, it seems to try to send audio buffer after (!) device was closed. This would be totally unrelated to the changes here though.

zetok commented:
~~I couldn't reproduce the segfault on master, while it's reproducible on this PR.~~

zetok commented:
Err, nvm, I've looked at the wrong terminal window, it's actually reproducible on master – PR is IMO good to go.

antis81 commented:
> … PR is IMO good to go.

Then please merge it quick and don't ask for long… :smile:

zetok commented:
> Then please merge it quick and don't ask for long… :smile:

Um, thing is, I don't even know C++. Merging PRs without being able to evaluate what the code change does doesn't sound like something that one should do, so feel free to merge :)

antis81 commented:
@zetok Ok, that's a sane decision then. Never meant to push you btw! Now I even understand better, why situation is as it is and merge process takes its time. You know - it's about responsibility and currently my personal situation is "suboptimal" (others might state it's disastrous actually :smile:) - IOW: My time on qTox is very limited and focused on pure development, code design & a bit review. Due to this, I am not able taking place in regular Saturday's online meetings. Anyway I'm convinced qTox will become the best (encrypted) messenger available and like to support the project. So… nvmd, I'll merge it myself.

zetok commented:
I see. It's really awesome that you find even a little bit of time for qTox even though things aren't going well well for you.

Hopefully things will get better for you soon :)

Btw, meetings lately haven't been really happening due to almost everyone not being able to participate, so don't worry about it :P
